jans-cedarling/cedarling/src/common/policy_store/entity_parser.rs

```diff
@@ -0,0 +1,686 @@
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+use std::collections::{HashMap, HashSet};
+use std::str::FromStr;
+
+/// A parsed Cedar entity with metadata.
+///
+/// Contains the Cedar entity and metadata about the source file.
+#[derive(Debug, Clone)]
+pub struct ParsedEntity {
+    /// The Cedar entity
+    pub entity: Entity,
+    /// The entity's UID
+    pub uid: EntityUid,
+    /// Source filename
+    pub filename: String,
+    /// Raw entity content (JSON)
+    pub content: String,
+}
+
+/// Raw entity JSON structure as expected by Cedar.
+///
+/// This matches Cedar's JSON entity format with uid, attrs, and parents fields.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RawEntityJson {
+    /// Entity unique identifier
+    pub uid: EntityUidJson,
+    /// Entity attributes as a map of attribute names to values (optional)
+    #[serde(default)]
+    pub attrs: HashMap<String, JsonValue>,
+    /// Parent entity UIDs for hierarchy (optional)
+    #[serde(default)]
+    pub parents: Vec<EntityUidJson>,
+}
+
+/// Entity UID in JSON format.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct EntityUidJson {
+    /// Entity type (e.g., "Jans::User")
+    #[serde(rename = "type")]
+    pub entity_type: String,
+    /// Entity ID
+    pub id: String,
+}
+
+/// Entity parser for loading and validating Cedar entities.
+pub struct EntityParser;
+
+impl EntityParser {
+    /// Parse a single Cedar entity from JSON content.
+    ///
+    /// Validates the entity structure, parses the UID, attributes, and parent relationships.
+    /// Optionally validates entity attributes against a schema.
+    ///
+    /// # Errors
+    /// Returns `PolicyStoreError` if:
+    /// - JSON parsing fails
+    /// - Entity structure is invalid
+    /// - UID format is invalid
+    /// - Parent UID format is invalid
+    /// - Schema validation fails (if schema provided)
+    pub fn parse_entity(
+        content: &str,
+        filename: &str,
+        schema: Option<&Schema>,
+    ) -> Result<ParsedEntity, PolicyStoreError> {
+        let json_value: JsonValue =
+            serde_json::from_str(content).map_err(|e| PolicyStoreError::JsonParsing {
+                file: filename.to_string(),
+                source: e,
+            })?;
+
+        // Parse the JSON structure
+        let raw_entity: RawEntityJson =
+            serde_json::from_value(json_value).map_err(|e| PolicyStoreError::JsonParsing {
+                file: filename.to_string(),
+                source: e,
+            })?;
+
+        // Parse the entity UID
+        let uid = Self::parse_entity_uid(&raw_entity.uid, filename)?;
+
+        // Parse parent UIDs (for hierarchy validation only)
+        let _parents: HashSet<EntityUid> = raw_entity
+            .parents
+            .iter()
+            .map(|parent_uid| Self::parse_entity_uid(parent_uid, filename))
+            .collect::<Result<HashSet<_>, _>>()?;
```

Comment:
If we don't need result `HashSet` I think no reason to collect values.

```suggestion
        for parent_uid in raw_entity.parents {
            Self::parse_entity_uid(parent_uid, filename)?
        }
```